### ISSUE TITLE
opt: panic: inner-join expression cannot be added to multiple groups: inner-join

### ISSUE THREAD

**Issue by jordanlewis** [cockroachdb/cockroach#35253]
```
CREATE TABLE table0 (col0 VARBIT, col1 STRING NOT NULL, col2 INET, col3 INT8, col4 INTERVAL NULL, col5 INT8[], col6 UUID NULL, col7 OID[], col8 INTERVAL, col9 UUID, col10 STRING, col11 INT8 NOT NULL, col12 BOOL NULL, col13 TIME, col14 BOOL NOT NULL, col15 NAME, col16 INET NOT NULL, FAMILY fam0 (col9, col1, col12, col2, col14, col11), FAMILY fam1 (col10, col0, col3), FAMILY fam2 (col5), FAMILY fam3 (col16), FAMILY fam4 (col4, col15, col8, col13, col7), FAMILY fam5 (col6), UNIQUE (col16 DESC), INDEX (col14), INDEX (col16, col10 ASC), INDEX (col16 DESC), INDEX (col9 DESC, col16 DESC, col1 DESC, col8 ASC, col0, col4, col12, col14, col2 ASC, col15 DESC, col3 DESC, col11 DESC), UNIQUE (col6 ASC, col14 ASC, col3 DESC, col16 DESC, col0 ASC));CREATE TABLE table1 (col0 TIMESTAMP NOT NULL, col1 TIMESTAMP NULL, col2 DECIMAL, col3 OID[], col4 UUID, col5 TIMESTAMP, col6 OID[] NOT NULL, col7 STRING COLLATE en NOT NULL, FAMILY fam0 (col6, col3, col4, col5, col7, col1, col2, col0), PRIMARY KEY (col7 DESC), INDEX (col1), INDEX (col0), INDEX (col2 DESC));CREATE TABLE table2 (col0 STRING, col1 STRING, col2 STRING COLLATE de NULL, col3 FLOAT8 NOT NULL, col4 INT8 NOT NULL, col5 STRING COLLATE en, col6 INT8, col7 TIMESTAMP NULL, col8 TIMESTAMPTZ NULL, col9 DECIMAL NULL, col10 UUID NOT NULL, FAMILY fam0 (col0, col9, col8), FAMILY fam1 (col4), FAMILY fam2 (col1, col2, col10, col7, col3), FAMILY fam3 (col5, col6), PRIMARY KEY (col9 DESC, col8, col0 ASC), INDEX (col4 ASC, col7 ASC, col3 DESC, col5 DESC, col9 ASC, col0 DESC, col2 ASC, col10, col6 DESC), INDEX (col0 DESC, col8 ASC, col7 DESC, col1 DESC, col3, col6 DESC, col10 ASC, col2 DESC, col5), INDEX (col5), UNIQUE (col3 DESC, col4 ASC, col2, col1 ASC, col8, col6, col0, col10 DESC, col5, col7 DESC, col9 ASC), UNIQUE (col8 ASC, col9, col0, col5 DESC, col4 DESC, col3, col2), UNIQUE (col8 DESC, col6, col4 ASC, col2, col10 ASC), UNIQUE (col4, col1 DESC, col8 DESC, col10 ASC, col3, col5 ASC), INDEX (col2, col3 ASC, col4 DESC, col10, col7 DESC, col9, col1 ASC), UNIQUE (col6, col5 DESC, col8, col9, col2 DESC, col3));CREATE TABLE table3 (col0 BOOL, col1 INTERVAL NULL, col2 STRING COLLATE de, col3 BOOL, col4 TIME NOT NULL, col5 UUID NOT NULL, col6 FLOAT8, col7 TIMESTAMP NULL, col8 DATE NULL, FAMILY fam0 (col3), FAMILY fam1 (col4, col2), FAMILY fam2 (col6), FAMILY fam3 (col5, col8), FAMILY fam4 (col1, col7, col0), PRIMARY KEY (col3 ASC), INDEX (col5 ASC, col6, col8 DESC, col7, col3, col2 ASC, col1 ASC, col4 ASC), UNIQUE (col5 DESC, col4 ASC), INDEX (col6, col8 DESC, col2 DESC, col0, col5 ASC, col1 ASC, col3 ASC), INDEX (col0, col6, col8 DESC, col5 ASC, col4, col1 ASC, col7 ASC, col3 ASC));CREATE TABLE table4 (col0 TIME NULL, col1 DATE, col2 BYTES NULL, col3 INTERVAL, col4 TIMESTAMP, col5 VARBIT NOT NULL, col6 INT8, col7 TIMESTAMPTZ NULL, col8 DECIMAL NOT NULL, col9 TIMESTAMPTZ, col10 OID[] NOT NULL, col11 TIMESTAMPTZ NOT NULL, col12 STRING COLLATE en, col13 OID NULL, FAMILY fam0 (col0, col5, col6, col11, col10), FAMILY fam1 (col3), FAMILY fam2 (col1, col2, col12, col13, col8, col7, col4, col9), PRIMARY KEY (col5, col11 ASC, col6 ASC), UNIQUE (col8, col1 ASC, col9 DESC, col3, col2, col5, col4, col7 DESC, col13 DESC, col0 DESC, col12 DESC, col11 DESC, col6 ASC), INDEX (col0, col1), INDEX (col4 DESC, col3));CREATE TABLE table5 (col0 TIME NULL, col1 STRING NOT NULL, col2 TIMESTAMPTZ, col3 FLOAT8 NOT NULL, col4 TIMESTAMPTZ NOT NULL, col5 VARBIT, col6 INTERVAL NOT NULL, col7 TIMESTAMPTZ NULL, col8 OID[] NULL, col9 DATE NOT NULL, col10 DATE NOT NULL, col11 TIMESTAMPTZ, col12 NAME, col13 TIMESTAMP NULL, col14 INT8[] NULL, col15 TIMESTAMPTZ NULL, col16 DATE NULL, col17 INT8 NULL, col18 TIMESTAMPTZ NOT NULL, FAMILY fam0 (col4), FAMILY fam1 (col0), FAMILY fam2 (col8), FAMILY fam3 (col13), FAMILY fam4 (col11), FAMILY fam5 (col14, col17), FAMILY fam6 (col6), FAMILY fam7 (col9, col12), FAMILY fam8 (col10), FAMILY fam9 (col16), FAMILY fam10 (col2), FAMILY fam11 (col18), FAMILY fam12 (col7), FAMILY fam13 (col1), FAMILY fam14 (col5), FAMILY fam15 (col15), FAMILY fam16 (col3), PRIMARY KEY (col4), UNIQUE (col15 ASC, col13, col7 ASC, col17 DESC, col0 DESC, col10 DESC, col5 ASC), UNIQUE (col17 DESC, col3 ASC, col13, col9, col15 DESC, col7), INDEX (col15 DESC, col0 DESC, col7 ASC, col10 DESC, col13 ASC, col12, col9 DESC, col5, col1 ASC), INDEX (col18 DESC, col10 ASC, col9 DESC, col2, col7 DESC, col3, col11, col6 ASC, col12), UNIQUE (col6 ASC, col18 ASC, col10, col16 DESC, col13, col1 ASC));CREATE TABLE table6 (col0 STRING NOT NULL, col1 INTERVAL NULL, col2 UUID, col3 INET NOT NULL, col4 INT8 NULL, col5 OID[] NULL, col6 INET NOT NULL, col7 BYTES NOT NULL, col8 STRING NOT NULL, col9 INT8, col10 BOOL NOT NULL, col11 TIMESTAMP NULL, col12 INT8[] NOT NULL, col13 DECIMAL NOT NULL, col14 INT8 NULL, col15 OID NOT NULL, col16 TIMESTAMP NULL, col17 OID NOT NULL, FAMILY fam0 (col5), FAMILY fam1 (col6), FAMILY fam2 (col1), FAMILY fam3 (col9, col13), FAMILY fam4 (col17, col2), FAMILY fam5 (col12), FAMILY fam6 (col10, col7), FAMILY fam7 (col4), FAMILY fam8 (col8, col14), FAMILY fam9 (col15), FAMILY fam10 (col16), FAMILY fam11 (col0, col11, col3), INDEX (col13, col9 DESC, col2 DESC, col6 ASC, col14, col7 ASC, col0, col1), INDEX (col8 DESC, col16 DESC, col6 DESC, col14, col11 ASC, col2, col1 DESC, col0 ASC, col17 ASC, col9 ASC, col13 DESC, col10, col4 ASC, col15, col3 DESC), UNIQUE (col6, col0 ASC, col9 ASC, col10 ASC, col3 ASC, col7 DESC, col4, col17 DESC, col1), UNIQUE (col2 DESC, col13 DESC, col1 DESC, col4, col17 ASC, col8, col16 ASC, col9, col14 DESC, col10 DESC, col11, col0 ASC, col6 DESC, col15 DESC), UNIQUE (col17 ASC, col0 ASC, col4 ASC, col6 DESC, col10, col15 ASC, col2 ASC, col16), UNIQUE (col0, col13 ASC, col17 DESC, col16 ASC, col1 DESC, col3, col2, col11 ASC, col9 DESC, col15 ASC, col7 ASC, col4 ASC, col14, col8 ASC), UNIQUE (col13 ASC, col7 DESC, col17, col6 DESC, col10, col3, col16, col14, col15));CREATE TABLE table7 (col0 BOOL, col1 BYTES NULL, FAMILY fam0 (col0), FAMILY fam1 (col1), PRIMARY KEY (col0 DESC), UNIQUE (col1 ASC, col0 DESC), UNIQUE (col1, col0 DESC), INDEX (col0 DESC), INDEX (col0, col1 DESC), UNIQUE (col1 ASC, col0), INDEX (col1 DESC));CREATE TABLE table8 (col0 DATE NULL, col1 OID[] NOT NULL, col2 VARBIT NOT NULL, col3 TIMESTAMP NOT NULL, col4 BOOL NULL, col5 TIMESTAMPTZ NULL, col6 INT8[], col7 TIME, col8 INT8[] NOT NULL, col9 FLOAT8, col10 DATE NULL, col11 DECIMAL, col12 BYTES NOT NULL, col13 BOOL, col14 VARBIT, col15 OID NULL, FAMILY fam0 (col4, col9, col11), FAMILY fam1 (col12, col7, col14, col6, col10, col8, col15, col3, col0), FAMILY fam2 (col5, col1), FAMILY fam3 (col2), FAMILY fam4 (col13), PRIMARY KEY (col9 ASC, col11 ASC, col4), UNIQUE (col12 DESC), UNIQUE (col10 DESC, col7, col9, col0 ASC, col14 ASC, col3), INDEX (col3));CREATE TABLE table9 (col0 TIMESTAMP NULL, col1 INET NOT NULL, col2 OID, FAMILY fam0 (col2, col1), FAMILY fam1 (col0), PRIMARY KEY (col2, col1 DESC), UNIQUE (col1 ASC, col0 DESC, col2), INDEX (col1, col0 DESC, col2));

select -0.8959352150683553 from [insert into table0 (rowid, col1, col3, col5, col6, col10, col11, col14, col15, col16) values (NULL, bpcharin('82.166.224.145/2'), NULL, ARRAY[2260128618786834763,-4098578038916670282,-4324793131055183163], '07b42336-2fda-4390-b125-04dcd271ef75', '6!Z3N', div(NULL, 2442412142335111980), false, e'\x13', '49a5:64a0:df6f:21be:b0f1:8298:378e:4637/1'), (case when true then 6742336964933357517 else NULL end, e'\x17', -33448820141811775, ARRAY[-7253305958742039619], '1ce9b3ae-eb5d-4fb5-beaa-44ad1d2c9cd8', e'L\'#rwCq', 7987498683458613158, true, namerecv('1978-12-15'), '7c0c:781e:4672:5fcb:bdde:db33:1941:b542/11'), (6284350805598536804, 'v"', case when true then NULL else 6017811722447645157 end, ARRAY[8465353088154290023], 'ba6ca413-84b7-4109-a68a-0ad2a54587b2', NULL, -211951226940815471, true, case when NULL then e'\x01k99R.' else e'=,\x1dV' end, '5280:639e:97e6:b089:2ad9:78fe:c036:7ffa/104') returning case when cast(coalesce(table0.col14, false) as boolean) then '1970-01-04 08:37:34.000822+00:00' else '1970-01-07 20:02:23.00089+00:00' end, case when table0.col12 then cast(coalesce(table0.col2, table0.col2) as inet) else NULL end, case when table0.col12 then '<vYx,' else e'c:m\x02\x12\x18' end, case when table0.col12 then (table0.col3 + -336650547691934603.1) else -5.766715185599382587E+33 end, e'\x7f.\x1aJ`l\rc', case when true then e'\x06\x15d d/\x0f0' else e'\x1e{[\x182\b\x10' end, e'{"KPW}hyA$\\"q$o": [], "c": "b", "foo": [false, {}, {}]}', 2227355386, table0.col8] join table3 as tab_2 join [insert into table0 (rowid, col1, col4, col5, col6, col7, col8, col9, col10, col11, col14, col15, col16, col0) select NULL, e':8z\x7fR6\x11', '-55 years -2 mons -557 days -20:52:11.803219', NULL, 'c7bcee57-a43b-4817-a05d-ce9b24a05519', NULL, '5 years 6 mons 872 days 06:06:24.139316', 'b445d05d-bc6e-4b98-ad2b-d1a80f301180', e'\x10w@\x0e', -7192100822834670366, false, 'tHW', 'a570:2eaa:e238:da61:52a0:e0d8:8015:d451/110', B'0100011110110111101101010001001011011' from table0 as tab_4 where true returning B'10001111110001', 8862811343975736589] on false join table5 as tab_5 on (e'\x19YY{\nf' NOT ILIKE e'BE=I0sS\rl') join [insert into table9 (col2, col1) select 543747490, '10.247.186.215/26' from table1 as tab_7 returning e',\x1aWa'] on (1208481304781103317 > 4.005544191387400841E+24) on (e'[{")^+9^|`_&#cJ": ["q3a>d7"], "OUW": {"Xy.YG+\'KVV3L": {}, "a": null, "foobar": {}}, "foo": {}}, {}, {"6-|0Q": null, "foobar": []}, [], {}]' IS DISTINCT FROM '[{"baz": [[[]], [[]], false, null], "c": [{}]}, {}]') where (NULL IS NOT DISTINCT FROM '8266e050-9bb9-445d-8c2e-bebe9885f265') limit 60;

E190228 00:29:05.123847 389 util/log/crash_reporting.go:202  [n1,client=127.0.0.1:54865,user=root] a panic has occurred!
E190228 00:29:05.129151 389 util/log/crash_reporting.go:476  [n1,client=127.0.0.1:54865,user=root] Reported as error 5ddc17b0bd3d479394290e606c9bb897
panic: inner-join expression cannot be added to multiple groups: inner-join
 ├── columns: table9.col0:127(timestamp) table9.col1:128(inet!null) table9.col2:129(oid!null)
 ├── cardinality: [0 - 0]
 ├── side-effects, mutations
 ├── stats: [rows=0]
 ├── fd: ()-->(127-129)
 ├── values
 │    ├── cardinality: [0 - 0]
 │    ├── stats: [rows=0]
 │    └── key: ()
 ├── inner-join
 │    ├── columns: table9.col0:127(timestamp) table9.col1:128(inet!null) table9.col2:129(oid!null)
 │    ├── cardinality: [0 - 0]
 │    ├── side-effects, mutations
 │    ├── stats: [rows=0]
 │    ├── fd: ()-->(127-129)
 │    ├── join-size: 2
 │    ├── values
 │    │    ├── cardinality: [0 - 0]
 │    │    ├── stats: [rows=0]
 │    │    └── key: ()
 │    ├── select
 │    │    ├── columns: table9.col0:127(timestamp) table9.col1:128(inet!null) table9.col2:129(oid!null)
 │    │    ├── cardinality: [0 - 0]
 │    │    ├── side-effects, mutations
 │    │    ├── stats: [rows=0]
 │    │    ├── fd: ()-->(127-129)
 │    │    ├── insert table9
 │    │    │    ├── columns: table9.col0:127(timestamp) table9.col1:128(inet!null) table9.col2:129(oid!null)
 │    │    │    ├── insert-mapping:
 │    │    │    │    ├──  column140:140 => table9.col0:127
 │    │    │    │    ├──  "?column?":139 => table9.col1:128
 │    │    │    │    └──  "?column?":138 => table9.col2:129
 │    │    │    ├── side-effects, mutations
 │    │    │    ├── stats: [rows=1000]
 │    │    │    ├── fd: ()-->(127-129)
 │    │    │    └── project
 │    │    │         ├── columns: column140:140(timestamp) "?column?":138(oid!null) "?column?":139(inet!null)
 │    │    │         ├── stats: [rows=1000]
 │    │    │         ├── fd: ()-->(138-140)
 │    │    │         ├── prune: (138-140)
 │    │    │         ├── scan tab_7
 │    │    │         │    └── stats: [rows=1000]
 │    │    │         └── projections
 │    │    │              ├── null [type=timestamp]
 │    │    │              ├── const: 543747490 [type=oid]
 │    │    │              └── const: '10.247.186.215/26' [type=inet]
 │    │    └── filters
 │    │         └── false [type=bool]
 │    └── filters (true)
 └── filters (true)
```

**Comment by andy-kimball** [comment]
Justin, another group merge to investigate...

**Comment by justinj** [comment]
I'll have to look more later, but here's a reduced repro:
```
exec-ddl
CREATE TABLE x (a INT8 PRIMARY KEY)
----
TABLE x
 ├── a int not null
 └── INDEX primary
      └── a int not null

opt join-limit=4 format=show-all
SELECT
    1
FROM
    [INSERT INTO x (a) VALUES (NULL) RETURNING 1]
    JOIN
        x AS y
        JOIN [INSERT INTO x (a) SELECT NULL FROM x RETURNING 1] ON false
        JOIN x ON true
        JOIN [INSERT INTO x (a) SELECT NULL FROM x RETURNING 1] ON true
    ON true
----
```

Not really sure what's going on yet and why it seems to depend on side-effects, since the only thing the side effects were doing before was preventing groups from collapsing into empty `VALUES` expressions and messing with the exploration order, but that should have stopped being an issue when we gave all the `VALUES` clauses unique IDs.

**Comment by justinj** [comment]
Fixed by #35482.
